### PR TITLE
Fully internalise balancing, fix tests

### PR DESCRIPTION
**Pull request by samuelWilliams99**
This interalises ada balancing, no longer relying on `cardano-cli transaction build`, and always building raw.
This means the Tx returned from `submitTx` or `balanceTx` is identical to the one submitted to the chain.
It also tweaks the balancing loop a bit to be more correct.
As such, `PreBalance` has been renamed to `Balance`

In order to handle the now changing txid of transaction as they are balanced, the tests were changed to now support a "?" in assertCommandHistory. Anywhere where a "?" is used in an expected command is seen as identical to up to the next space in the actual command.

See it as a regex `^[^ ]+(?= )`.
e.g.
```
this is an example actual command
```
As an actual command would compare equal to:
```
this is an example ? command
```

### CODE REVIEW

File: test/Spec/BotPlutusInterface/Balance.hs [hunk 1]
```diff
@@ -97,11 +87,8 @@ addUtxosForChange = do
...
-      requiredSigs = [pkh1]
       ownPkh = pkh1
-      pparams = def {protocolParamUTxOCostPerWord = Just (Lovelace 1)}
-      prebalancedTx =
-        PreBalance.preBalanceTx pparams minUtxo fees utxoIndex ownPkh privKeys requiredSigs tx
+      balancedTx =
+        Balance.balanceTxStep minUtxo fees utxoIndex ownPkh tx
 
-  txInputs <$> prebalancedTx @?= Right (Set.fromList [txIn1, txIn2, txIn3])
+  txInputs <$> balancedTx @?= Right (Set.fromList [txIn1, txIn2])
```
Comment (samuelWilliams99) [inline test/Spec/BotPlutusInterface/Balance.hs:94]: Note this input was removed, unsure why it was ever there?
The expected output of this tx is 1.6ADA + 0.5ADA fee, inputs 1 and 2 are 1.1ADA and 1.0ADA, both summing to 2.1ADA, so only 2 inputs should be needed to cover it.

File: test/Spec/BotPlutusInterface/Contract.hs [hunk 2]
```diff
@@ -129,38 +130,41 @@ sendAda = do
...
+          --tx-out-count 2
           --witness-count 1
           --protocol-params-file ./protocol.json
           --mainnet
           |]
         )
-      ,
-        ( 6
+      , -- Steps 4, 5 and 6 are near repeats of 1, 2 and 3, to ensure min utxo values are met
+
```
Comment (samuelWilliams99) [inline test/Spec/BotPlutusInterface/Contract.hs:154]: Fourmolu decided this space should be here, not sure why

File: test/Spec/BotPlutusInterface/Contract.hs [hunk 3]
```diff
@@ -851,9 +870,27 @@ assertCommandHistory :: forall (w :: Type). MockContractState w -> [(Int, Text)]
...
+commandEqual :: Text -> Text -> Bool
+commandEqual "" "" = True
+commandEqual "" _ = False
+commandEqual _ "" = False
+commandEqual expected actual = maybe False (on commandEqual dropToSpace postExp) mPostAct
+  where
+    (preExp, postExp) = Text.breakOn "?" expected
+    mPostAct = Text.stripPrefix preExp actual
+    dropToSpace = Text.dropWhile (/= ' ')
```
Comment (samuelWilliams99) [inline test/Spec/BotPlutusInterface/Contract.hs:892]: This could likely be done better, didn't want to bring all of regex in though, any ideas?

File: test/Spec/BotPlutusInterface/Balance.hs (hunk 1 above)
Comment (gege251) [reply test/Spec/BotPlutusInterface/Balance.hs:94]: I think this is the edge case we talked about, when even though the inputs and outputs are balanced, the cli adds a change utxo which must be at least the minUtxo (around 1 ada). This is now outdated.
I'm wondering though what happens if we change the amout being sent to 1.5, so the whole amount is 2ADA, with a change of 0.1, which then must be rounded up.

File: test/Spec/BotPlutusInterface/Contract.hs (hunk 3 above)
Comment (gege251) [reply test/Spec/BotPlutusInterface/Contract.hs:892]: I can't come up with anything better right how either.
However, I think we could have the trimming of newlines on line 875 merged into this function

**Review by gege251 — COMMENTED** [review]
Nice to see that actually the code got simpler

File: src/BotPlutusInterface/Contract.hs [hunk 4]
```diff
@@ -163,15 +163,15 @@ handlePABReq contractEnv req = do
...
 -- | This is not identical to the real balancing, we only do a pre-balance at this stage
-balanceTx ::
+balanceTxStep ::
   forall (w :: Type) (effs :: [Type -> Type]).
   Member (PABEffect w) effs =>
   ContractEnvironment w ->
   UnbalancedTx ->
   Eff effs BalanceTxResponse
-balanceTx contractEnv unbalancedTx = do
+balanceTxStep contractEnv unbalancedTx = do
```
[the remark this answers is not in the document]
Comment (samuelWilliams99) [reply src/BotPlutusInterface/Contract.hs:172]: This can add an extra output, in which case the result isnt fully balanced, and as such, another step would be needed

File: test/Spec/BotPlutusInterface/Contract.hs (hunk 3 above)
Comment (samuelWilliams99) [reply test/Spec/BotPlutusInterface/Contract.hs:892]: Oh yeah, good point

**Comment by samuelWilliams99** [comment]
Following a real world test, I've updated the tests to now scale fee as inputs/outputs update. This exposed a fault with the ada balancing, I've fixed this with some unfortunately not super beautiful code.

File: src/BotPlutusInterface/Contract.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (samuelWilliams99) [reply src/BotPlutusInterface/Contract.hs:172]: Oh never mind! It was a project wide rename, and `balanceTx` was used multiple times

File: src/BotPlutusInterface/Balance.hs [hunk 5]
```diff
@@ -146,7 +168,26 @@ balanceTxStep ::
 balanceTxStep minUtxos fees utxos ownPkh tx =
```
Comment (gege251) [inline src/BotPlutusInterface/Balance.hs:168]: Could we add fees to directly to Tx and remove the fee from the arguments (just a suggestion)?

**Comment by gege251** [comment]
balancing is hard...

File: src/BotPlutusInterface/Balance.hs [hunk 6]
```diff
@@ -83,16 +81,45 @@ preBalanceTxIO pabConf ownPkh unbalancedTx =
...
+
+      -- Balance the tx
+      (balancedTx, minUtxos) <- loop utxoIndex privKeys [] preBalancedTx
+
+      -- Check if we have Ada change
+      let adaChange = getAdaChange utxoIndex (lovelaceValue $ txFee balancedTx) balancedTx
+      -- If we have no change UTxO, but we do have change, we need to add an output for it
+      -- We'll add a minimal output, run the loop again so it gets minUTxO, then update change
+      balancedTxWithChange <-
+        if adaChange /= 0 && not (hasChangeUTxO ownPkh balancedTx)
```
Comment (gege251) [inline src/BotPlutusInterface/Balance.hs:96]: It's got a bit complicated here, I wonder if we could somehow refactor this. But for that, maybe it's best if we create some property tests for this.. which could be a bit laborious as we we need to simulate the calculateFee and calculateMinUtxo behaviours somehow, so it might be best if we do it in a later PR.

**Review by gege251 — APPROVED** [review]
TBH, I'm happy with this as it is, as refactoring balancing might not worth the trouble for now.

File: src/BotPlutusInterface/Balance.hs (hunk 5 above)
Comment (samuelWilliams99) [reply src/BotPlutusInterface/Balance.hs:168]: I actually considered this then got pulled into a meeting, will do!

File: src/BotPlutusInterface/Balance.hs (hunk 6 above)
Comment (samuelWilliams99) [reply src/BotPlutusInterface/Balance.hs:96]: I'll try refactor it a bit, but yeah, the tests are a bit lacking, simply due to the intended behavior being complex and somewhat hidden

**Comment by samuelWilliams99** [comment]
Tested with the transfer example - all good.
